### ISSUE TITLE
Memory Leak Using Plotter and Screenshot

### ISSUE THREAD

**Issue by AnvarAtayev** [pyvista/pyvista#2252]
Memory Leak Using Plotter and Screenshot
-----
It seems that `Plotter.screenshot()` produces a memory leak. I am currently running an iterative program implementing the observer pattern, where after each iteration (i.e. when the observer is notified), the observer plots and saves (as a .png) OpenFOAM data.

What I notice in my Activity Monitor is that after each time the observer is called and executed (i.e. PlotVTK.update() is run), my python memory increases by around  ~70MB, which becomes a problem when the observer is called around 300 times, when memory consumption has increased to many many GBs.

I have noticed that when I comment out `p.screenshot(f'plot_{iteration}', window_size=[1000,1000], return_img = False)
`, my memory increase after each observer call is much smaller, i.e. 5MB. Although, when I don't call this observer at all, my memory increase after each observer call in <1MB.

Any help would be appreciated.

**My code**

```py
import Observer
import pyvista as pv

class PlotVTK(Observer):
    
    def update(self, subject) -> None:
        
        # Get current iteration
        iteration = subject.iteration
        
        # Open openfoam file
        reader = pv.OpenFOAMReader(f'test_{iteration}.foam')

        # Obtain data for latest snapshot
        reader.set_active_time_value(reader.time_values[-1])

        # Read data into mesh
        mesh = reader.read()

        # Extract internal mesh data
        internalMesh = mesh["internalMesh"].slice_orthogonal()

        # Create plotting environment
        p = pv.Plotter(off_screen = True)

        # Plot velocity diagram
        p.add_mesh(internalMesh, scalars = 'U')

        # Set viewing angle
        p.view_xy()
        p.camera_position = ((0, 0, 3), (0, 0, 0), (0, 0, 0))
        
        # Take screenshot of image and save in appropriate directory
        p.screenshot(f'plot_{iteration}', window_size=[1000,1000], return_img = False)

        # Clear plot data (i.e. release memory)
        p.clear()
```
-----

**System Information:**
```txt

--------------------------------------------------------------------------------
  Date: Mon Feb 21 16:38:57 2022 GMT

                OS : Darwin
            CPU(s) : 8
           Machine : x86_64
      Architecture : 64bit
               RAM : 16.0 GiB
       Environment : IPython
       File system : apfs
        GPU Vendor : Apple
      GPU Renderer : Apple M1
       GPU Version : 4.1 Metal - 76.3

  Python 3.9.7 (default, Sep 16 2021, 08:50:36)  [Clang 10.0.0 ]

           pyvista : 0.33.2
               vtk : 9.1.0
             numpy : 1.21.2
           imageio : 2.16.0
           appdirs : 1.4.4
            scooby : 0.5.11
        matplotlib : 3.5.0
           IPython : 7.29.0
        ipyvtklink : 0.2.2
             scipy : 1.7.3
              tqdm : 4.62.3
--------------------------------------------------------------------------------

```

**Comment by RichardScottOZ** [comment]
Is the saved img size approx that memory leak difference?

**Comment by AnvarAtayev** [comment]
@RichardScottOZ Interestingly, no, the image files that are produced are only around 100KB in size, where as the memory leak is up to 70MB. Not sure what is going on with it. I imagine the same behaviour can be replicated if the above is put in a for loop.

**Comment by RichardScottOZ** [comment]
Yes, strange! I was doing a bunch with screenshots the other day and things were getting slow...but I wasn't monitoring memory.

**Comment by RichardScottOZ** [comment]
Yes, I am seeing this now, had a loop of about 3000 and deleting the objects at the end of the loop and it is still climbing - basically doing the same as this, so there is definitely a problem.

**Comment by RichardScottOZ** [comment]
So I suppose next thing to do is profile it and try and find out where.

**Comment by akaszynski** [comment]
Could you please try [pyvista.Plotter.deep_clean](https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.deep_clean.html)?

**Comment by RichardScottOZ** [comment]
Looks like it gives me a 'can't set attribute' error?

```python

p = pv.Plotter(notebook=True)

p.add_mesh(mesh=thres, clim=[threshold,4], cmap=cmapCET, opacity=0.67)
p.show_grid(color='black')
p.set_background(color='brown')
p.screenshot(r'F:\BananaSplits.png')
#p.show()
p.deep_clean()

```

**Comment by akaszynski** [comment]
> Looks like it gives me a 'can't set attribute' error?

This is with 0.34?

**Comment by RichardScottOZ** [comment]
Yes, installed it yesterday.

**Comment by RichardScottOZ** [comment]
I will try a generic

**Comment by RichardScottOZ** [comment]
![image](https://user-images.githubusercontent.com/72196131/167518772-c58128d1-27ac-49c3-b13c-88b7604b4ba1.png)

**Comment by RichardScottOZ** [comment]
```python
grid = pv.UniformGrid()
grid.dimensions=np.array( [25, 25, 24] ) + 1
grid.origin= (100000,7000000,-3400)
grid.spacing = (200,200,200)

p = pv.Plotter(notebook=True)
p.add_mesh(mesh=grid)
p.show_grid(color='black')
p.set_background(color='brown')
p.screenshot(r'BananaSplits.png')
p.deep_clean()
```

**Comment by RichardScottOZ** [comment]
so maybe have to check those exist first?

**Comment by akaszynski** [comment]
That's a bug on our end. Can you clone and comment out 3199 and 3200 in `plotter.py`? I'm surprised that error hasn't come up in testing.

**Comment by RichardScottOZ** [comment]
I will try and give that a shot this afternoon, have to go out shortly.

**Comment by RichardScottOZ** [comment]
Still definitely leaking.

**Comment by factoryofthesun** [comment]
I have also been experiencing this issue on my Macbook M1. So far I've tried turning off the call to the `renderer.ResetCamera()` based off [this old discussion ](https://vtkusers.public.kitware.narkive.com/T0aFEBit/vtk-not-releasing-memory-including-minimal-example) as well as call the following after every `screenshot()` call but am still getting major memory leak. 

```
del pvmesh
del plotter.ren_win
plotter.close()
plotter.deep_clean()
pv.plotting._ALL_PLOTTERS.pop(plotter._id_name)
plotter.renderers = None
delattr(plotter, 'renderers')
plotter._theme = None
del plotter
pv.close_all()
```

**Comment by WesleyTheGeolien** [comment]
@AnvarAtayev @RichardScottOZ I wrote some code here: https://github.com/pyvista/pyvista/issues/2577#issue-1230912097

that checks object size before and after running the code, maybe you could run it in your cases to see of there are any objects that are lingering in python memory

**Comment by RichardScottOZ** [comment]
Thanks Wesley, will see if I can try it on the weekend.

**Comment by akaszynski** [comment]
This should be partially dealt with on the `main` branch. We've eliminated the need to deep clean and aside from dynamically creating classes (it's on the short todo list) we should have the memory issue taken care of. I'd appreciate if you could checkout `main` and let me know if it solves the issue for you as well.

**Comment by leo2r** [comment]
So I've been having the same problem for ages, and actually just managed to fix it now! I was creating a new Plotter instance in my method that was getting called at each step, and instead I took it out and made it a Class variable and just call ```pl.clear()``` at the start of the method. So my code now looks like this:

```python
class Env
    pl = pv.Plotter(off_screen=True)

    def get_image(self, cpos=[0, 0, 0], croll=0):
        self.pl.clear()
        self.pl.add_mesh(self.bone_mesh, color="lightblue", opacity=0.3) 
        self.pl.camera.position = cpos
        self.pl.camera.focal_point = (0,0,0)
        self.pl.camera.roll = croll
        img = self.pl.screenshot(None, return_img=True, window_size=(self.WINDOW_SIZE))
        img = img[..., :1]
        return img
```

It seems like there's no more memory hogging now and my code that used to slow down and freeze my computer at around 80 steps has run for well over 3000 steps now with no slowing down!

**Comment by chapochn** [comment]
@leo2r, if I understand your code, this is a workaround to reuse the same plotter object instead of creating a new one, right?

Any update on this issue? My code is still crashing because of this leak.

**Comment by leo2r** [comment]
@chapochn Yes exactly, that seemed to solve the memory issue/leak for me! Have you tried this workaround?

**Comment by chapochn** [comment]
@leo2r Actually I spoke too quickly. Your code didn't directly work in my case. I did pl.clear_actors() instead of pl.clear(), as the latter removes all the default plotting properties

**Comment by MouseAndKeyboard** [comment]
Also experiencing this!
